### PR TITLE
Davecaos.6.basic implentation for votes

### LINKED ISSUE #6: Implement basic functionality for Votes
Add the following public functions to `conferl` (with the associated test, documentation, repos and docs)

- [ ] ``vote_message(MessageId :: integer(), UserId :: integer(), Direction :: up | down) -> conferl_votes:vote().``
- [ ] ``unvote_message(MessageId :: integer(), UserId :: integer()) -> ok.``
- [ ] ``remove_message(VoteId :: integer()) -> ok.``
- [ ] ``fetch_vote(MessageId :: integer(), UserId :: integer()) -> notfound | conferl_votes:vote().``
- [ ] ``fetch_vote(VoteId :: integer()) -> notfound | conferl_votes:vote().``
- [ ] ``list_votes(MessageId :: integer()) -> [conferl_votes:vote()].``
- [ ] ``message_score(MessageId :: integer()) -> integer().``
- [ ] ``message_vote_counts(MessageId :: integer()) -> #{up | down => integer()}.``

### PR DESCRIPTION
**Pull request by davecaos**
Fix #6

### CODE REVIEW

File: test/cnf_vote_SUITE.erl [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+-spec test_counts_votes_message(config()) -> ok.
+test_counts_votes_message(_Config)  -> 
+  MessageId = 4,
+  CountVotesUp = 15,
+  CountVotesDown = 30,
+  [cnf_vote_repo:vote_message(UserId, MessageId) 
+    || UserId <- lists:seq(1,CountVotesUp)],
+  [cnf_vote_repo:unvote_message(UserId, MessageId) 
+    || UserId <- lists:seq(1,CountVotesDown)],
```
Comment (gadgetci) [inline test/cnf_vote_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 116

File: test/cnf_vote_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ok.
+
+
+-spec test_counts_votes_message(config()) -> ok.
+test_counts_votes_message(_Config)  -> 
+  MessageId = 4,
+  CountVotesUp = 15,
+  CountVotesDown = 30,
+  [cnf_vote_repo:vote_message(UserId, MessageId) 
+    || UserId <- lists:seq(1,CountVotesUp)],
```
Comment (gadgetci) [inline test/cnf_vote_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 114

File: src/cnf_vote.erl [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    , thumb      => Thumb
+   }.
+
+-spec id(vote()) -> integer().
+id(Vote) -> 
+  maps:get(id, Vote).
+
+-spec user_id(vote()) -> integer().
+user_id(Vote) -> 
+  maps:get(user_id,Vote).
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: missing space, again undetected by @elvisci. Waiting on inaka/elvis#221 :(

File: src/cnf_vote.erl [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec user_id(vote(), integer()) -> vote().
+user_id(Vote, UserId) -> 
+  Vote#{ user_id => UserId}.
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: remove space after {

File: src/cnf_vote.erl [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec message_id(vote()) -> integer().
+message_id(Vote) -> 
+  maps:get(message_id,Vote).
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: missing space

File: src/cnf_vote.erl [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec message_id(vote(), integer()) -> vote().
+message_id(Vote, UserId) -> 
+  Vote#{ message_id => UserId}.
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: remove space

File: src/cnf_vote.erl [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec thumb(vote()) -> thumb().
+thumb(Vote) -> 
+  maps:get(thumb, Vote).
+
+-spec thumb(vote(), thumb()) -> vote().
+thumb(Vote, Thumb) -> 
+  Vote#{ thumb => Thumb}.
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: remove space

File: src/cnf_vote.erl [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec thumb_wakeup(sumo:doc()) -> vote().
+thumb_wakeup(Vote = #{ thumb := 1}) ->
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: remove spaces after `#{`

File: src/cnf_vote.erl [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+thumb(Vote) -> 
+  maps:get(thumb, Vote).
+
+-spec thumb(vote(), thumb()) -> vote().
+thumb(Vote, Thumb) -> 
+  Vote#{ thumb => Thumb}.
+
+-spec thumb_wakeup(sumo:doc()) -> vote().
+thumb_wakeup(Vote = #{ thumb := 1}) -> 
+Vote#{ thumb => up};
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: indent

File: src/cnf_vote.erl [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+thumb_wakeup(Vote = #{ thumb := 0}) -> 
+Vote#{ thumb => down}.
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: indent

File: src/cnf_vote.erl [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+-spec thumb_sleep(vote()) -> sumo:doc().
+thumb_sleep(Vote = #{ thumb := up}) -> 
+Vote#{ thumb => 1};
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: indent

File: src/cnf_vote.erl [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+thumb_sleep(Vote = #{ thumb := down}) -> 
+Vote#{ thumb => 0}.
```
Comment (elbrujohalcon) [inline src/cnf_vote.erl]: indent

File: src/cnf_vote_repo.erl [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+% Unless required by applicable law or agreed to in writing,
+% software distributed under the License is distributed on an
+% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+% KIND, either express or implied.  See the License for the
+% specific language governing permissions and limitations
+% under the License.
+-module(cnf_vote_repo).
+
+-author('David Cao <david.cao@inakanetworks.com>').
+-export([ vote_message/2
```
Comment (elbrujohalcon) [inline src/cnf_vote_repo.erl]: let's change this functions to upvote and downvote… My original names were poorly chosen. Sorry about that.

File: src/cnf_vote_repo.erl [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+-spec remove_vote(integer()) -> boolean().
+remove_vote(VoteId) -> 
+  sumo:delete(cnf_vote, VoteId).
+
+-spec fetch_vote(integer(), integer()) -> notfound | cnf_vote:vote().
+fetch_vote(MessageId, UserId) -> 
+  Result = 
+    sumo:find_by(cnf_vote, [{message_id, MessageId}, {user_id, UserId}]),
+  case Result of
+    []     -> notfound;
```
Comment (elbrujohalcon) [inline src/cnf_vote_repo.erl]: In this case you're returning `notfound`. I thought your idea was to always trhow `not_found` instead.

File: src/cnf_vote_repo.erl [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+fetch_vote(VoteId) -> 
+  sumo:find(cnf_vote, VoteId).
+
+-spec list_votes(integer()) -> [cnf_vote:vote()].
+list_votes(MessageId) -> 
+  sumo:find_by(cnf_vote, [{message_id, MessageId}]).
+
+-spec inc_thumbs(cnf_vote:vote(), thumb_count()) -> 
+  #{up | down => integer()}.
+inc_thumbs(Vote, ThumbAccumulator) ->
```
Comment (elbrujohalcon) [inline src/cnf_vote_repo.erl]: not exported functions should go either at the end of the module or **after** the only exported function that uses them, not before.

File: src/cnf_vote_repo.erl [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  Thumb = cnf_vote:thumb(Vote),
+  maps:update(Thumb, maps:get(Thumb, ThumbAccumulator) + 1, ThumbAccumulator).
+
+-spec counts_votes_message(MessageId :: integer()) -> thumb_count().
+counts_votes_message(MessageId) ->
+  Votes = list_votes(MessageId),
+  lists:foldl(fun inc_thumbs/2, #{up =>0, down => 0}, Votes).
+
+-spec delete_all() -> non_neg_integer().
+delete_all() ->
```
Comment (elbrujohalcon) [inline src/cnf_vote_repo.erl]: When do we need this for?

File: test/cnf_vote_SUITE.erl [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+init_per_testcase(_Function, Config) -> 
+  Config.
+
+%% @doc definion of end_per_testcases
+end_per_testcase(_Function, Config) -> 
+  Config.
+
+-spec test_vote(config()) -> ok.
+test_vote(_Config) -> 
+UserId = 1,
```
Comment (elbrujohalcon) [inline test/cnf_vote_SUITE.erl]: indent

File: test/cnf_vote_SUITE.erl [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+MessageId = 1,
+VotedMessage = cnf_vote_repo:vote_message(UserId, MessageId),
+UserId = cnf_vote:user_id(VotedMessage),
+MessageId = cnf_vote:user_id(VotedMessage),
+up = cnf_vote:thumb(VotedMessage),
+ok.
+
+-spec test_unvote(config()) -> ok.
+test_unvote(_Config) -> 
+UserId = 2,
```
Comment (elbrujohalcon) [inline test/cnf_vote_SUITE.erl]: indent

File: test/cnf_vote_SUITE.erl [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+MessageId = 2,
+VotedMessage = cnf_vote_repo:unvote_message(UserId, MessageId),
+UserId = cnf_vote:user_id(VotedMessage),
+MessageId = cnf_vote:user_id(VotedMessage),
+down = cnf_vote:thumb(VotedMessage),
+ok.
+
+-spec test_fetch_vote(config()) -> ok.
+test_fetch_vote(_Config) -> 
+UserId = 3,
```
Comment (elbrujohalcon) [inline test/cnf_vote_SUITE.erl]: indent

**Pull request merged**